### PR TITLE
Support fixed ports in transports

### PR DESCRIPTION
**Pull request by nazar-pc**
This is a simple non-breaking straightforward implementation of fixed ports support on transports that covers my use case from https://mediasoup.discourse.group/t/port-in-transport-options/2937.

This may evolve/change in v4, but should be useful in the meantime nevertheless.

### CODE REVIEW

File: worker/src/RTC/PortManager.cpp [hunk 1]
```diff
@@ -314,6 +314,184 @@ namespace RTC
...
+				break;
+		}
+
+		// Try to bind on it.
+		switch (transport)
+		{
+			case Transport::UDP:
+				uvHandle = reinterpret_cast<uv_handle_t*>(new uv_udp_t());
+				err      = uv_udp_init_ex(
+          DepLibUV::GetLoop(), reinterpret_cast<uv_udp_t*>(uvHandle), UV_UDP_RECVMMSG);
```
Comment (jmillan) [inline worker/src/RTC/PortManager.cpp:392]:
```suggestion
				err      = uv_udp_init_ex(
                                DepLibUV::GetLoop(), reinterpret_cast<uv_udp_t*>(uvHandle), UV_UDP_RECVMMSG);
```

File: worker/include/RTC/TcpServer.hpp [hunk 2]
```diff
@@ -32,6 +34,7 @@ namespace RTC
 		// Passed by argument.
 		Listener* listener{ nullptr };
 		RTC::TcpConnection::Listener* connListener{ nullptr };
+		bool fixedPort;
```
Comment (jmillan) [inline worker/include/RTC/TcpServer.hpp]:
```suggestion
		bool fixedPort{ false };
```

File: worker/include/RTC/UdpSocket.hpp [hunk 3]
```diff
@@ -28,6 +29,7 @@ namespace RTC
 	private:
 		// Passed by argument.
 		Listener* listener{ nullptr };
+		bool fixedPort;
```
Comment (jmillan) [inline worker/include/RTC/UdpSocket.hpp]:
```suggestion
		bool fixedPort{ false };
```

**Review by jmillan — APPROVED** [review]
LGTM

File: worker/src/RTC/PortManager.cpp (hunk 1 above)
Comment (nazar-pc) [reply worker/src/RTC/PortManager.cpp:392]: Can't do this, linter is not happy in that case :slightly_frowning_face:

Comment (jmillan) [reply worker/src/RTC/PortManager.cpp:392]: ouch, not even setting each parameter on its own line?

Comment (nazar-pc) [reply worker/src/RTC/PortManager.cpp:392]: Yeah, it wants everything on one line with that amount of indentation :confused:

Comment (jmillan) [reply worker/src/RTC/PortManager.cpp:392]: :+1:

**Comment by ibc** [comment]
Sorry for the delay, pseudo days off. Let me a few days before I do proper review.

File: worker/src/RTC/TcpServer.cpp [hunk 4]
```diff
@@ -26,7 +35,10 @@ namespace RTC
 	{
 		MS_TRACE();
 
-		RTC::PortManager::UnbindTcp(this->localIp, this->localPort);
+		if (!fixedPort)
+		{
+			RTC::PortManager::UnbindTcp(this->localIp, this->localPort);
+		}
```
Comment (ibc) [inline worker/src/RTC/TcpServer.cpp:41]: @nazar-pc, here and in `UdpSocket.cpp` I don't see where the given fixed `port` is unbinded once socket is closed. Am I missing something?

Comment (ibc) [reply worker/src/RTC/TcpServer.cpp:41]: Forget it, such a port is unbinded by the UdpSocketHandler and so on. It's ok.

**Review by ibc — APPROVED** [review]

**Comment by ibc** [comment]
@nazar-pc sorry for late merging. On it. May I ask you to send a PR to mediasoup-website documenting the new `port` field?

**Comment by ibc** [comment]
```
  ● router.createPlainTransport() with fixed port succeeds

    uv_udp_bind() failed [transport:udp, ip:'127.0.0.1', port:60000: address already in use
```

```

gst-launc 31397             ibc   16u  IPv4 0xa555a913b1bb1a49      0t0    UDP *:60000
```

I'm gonna change the port otherwise I won't be able to run tests locally 😅

**Comment by ibc** [comment]
Released in 3.7.18!

**Comment by jmillan** [comment]
> I'm gonna change the port otherwise I won't be able to run tests locally 😅

Handled: https://github.com/versatica/mediasoup/pull/599

File: worker/src/RTC/PortManager.cpp [hunk 5]
```diff
@@ -314,6 +314,184 @@ namespace RTC
...
+				  reinterpret_cast<const struct sockaddr*>(&bindAddr),
+				  flags);
+
+				if (err != 0)
+				{
+					// If it failed, close the handle and check the reason.
+					uv_close(reinterpret_cast<uv_handle_t*>(uvHandle), static_cast<uv_close_cb>(onClose));
+
+					MS_THROW_ERROR(
+					  "uv_udp_bind() failed [transport:%s, ip:'%s', port:%" PRIu16 ": %s",
```
Comment (stefansundin) [inline worker/src/RTC/PortManager.cpp:432]: This line is missing a `]`.

Comment (nazar-pc) [reply worker/src/RTC/PortManager.cpp:432]: Would you mind submitting a PR with a fix?

Comment (stefansundin) [reply worker/src/RTC/PortManager.cpp:432]: https://github.com/versatica/mediasoup/pull/663

Thanks!

**Pull request merged**